### PR TITLE
Iss78: Adding support for acompcor compartments

### PR DESCRIPTION
**Pull request by smeisler**
#78

Adds 1) Support for acompcor compartments (WM+CSF combined or separate masks), 2) functionality for choosing all compcor components to explain 50% variance (see [Mushelli et al, 2014](https://www.sciencedirect.com/science/article/pii/S105381191400175X?casa_token=0XyyOvOjc_MAAAAA:vGPITamIIWOArAFhAt0NX4TfAvxkXhJUiDVMR6ayt_CmEyhpm0MYK-5D7YhO2UAQQ6ulaLLV6mws))

This required some pretty substantial changes to the` _label_compcor` function, and also introduces loading the corresponding confounds json file.

Also, I had the censoring code (see #99 ) in my master branch by accident before making a branch for this issue, which is why you may see a lot of deletions. Sorry for the sloppiness!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,5 @@
 # load_confounds
+Adding acompcor compartment selection and 50% variance.
```
Comment (pbellec) [inline README.md:2]: Please remove. I will add contributions with the all-contributors bot, and the release change logs. This contribution will not be forgotten :)

File: load_confounds/parser.py [hunk 2]
```diff
@@ -76,34 +77,37 @@ def _load_high_pass(confounds_raw):
...
-        else:
-            compcor_cols.append(compcor_col)
-
+    compcor_cols = [key for key in confounds_json.keys() if compcor_suffix+'_comp' in key]
+    if compcor_suffix == 'a':
+        if acompcor_mask == 'combined':
+            compcor_cols = [key for key in compcor_cols if confounds_json[key]['Mask']=='combined']
+        else: compcor_cols = [key for key in compcor_cols if confounds_json[key]['Mask']!='combined']
+    if n_compcor != None:
+        if n_compcor > np.size(compcor_cols):
```
Comment (pbellec) [inline load_confounds/parser.py:88]: This block could be turned into a small function  `_sanitize_n_compcor` called at the beginning. This will help readability.

File: load_confounds/parser.py [hunk 3]
```diff
@@ -76,34 +77,37 @@ def _load_high_pass(confounds_raw):
...
+            warnings.warn("Number of requested compcor components ("+str(n_compcor)+") exceeds number of components ("+str(np.size(compcor_cols))+").")
+            n_compcor = np.size(compcor_cols)
+        if compcor_suffix == 'a' and acompcor_mask != 'combined':
+            n_compcor = int(np.ceil(n_compcor/2))
```
Comment (pbellec) [inline load_confounds/parser.py:92]: this block could be combined with the `else` statement line 86

File: load_confounds/parser.py [hunk 4]
```diff
@@ -76,34 +77,37 @@ def _load_high_pass(confounds_raw):
...
+            csf_comps = [comp for comp in compcor_cols if confounds_json[comp]['Mask']=='CSF']
+            wm_comps = [comp for comp in compcor_cols if confounds_json[comp]['Mask']=='WM']
+            compcor_cols = csf_comps[0:n_compcor] + wm_comps[0:n_compcor]
+        else: compcor_cols = compcor_cols[0:n_compcor]
```
Comment (pbellec) [inline load_confounds/parser.py:96]: this could be turned into an `elif n_compcor != None` and get rid of line 87

File: load_confounds/parser.py [hunk 5]
```diff
@@ -76,34 +77,37 @@ def _load_high_pass(confounds_raw):
     high_pass_params = _find_confounds(confounds_raw, ["cosine"])
     return confounds_raw[high_pass_params]
 
-
-def _label_compcor(confounds_raw, compcor_suffix, n_compcor):
+def _label_compcor(confounds_json, compcor_suffix, n_compcor, acompcor_mask):
```
Comment (pbellec) [inline load_confounds/parser.py:80]: this function has many if statements, making the flow of execution very hard to follow. I made some suggestions to try and simplify.

**Review by pbellec — CHANGES_REQUESTED** [review]
Great contribution! 

There is one function which I think should be simplified. 

Also you may want to add more parameters to the AnatCompCor strategy, and add some test for the variants.

**Comment by pbellec** [comment]
not sure why but there is a conflict with master, likely from the merge of PR #96 

I also see that the new default for `n_compcor` is `None` instead of 6.  Does that mean that by default no compcor components would be included? that seems strange if the user indicated wanting to include compcor regressors already.

**Comment by smeisler** [comment]
The n_compcor default being none means it will collect all compcor components (50% variance) if compcor is enabled.

**Comment by pbellec** [comment]
> The n_compcor default being none means it will collect all compcor components (50% variance) if compcor is enabled.

which also means I got lost in the if ... then :smile:

**Comment by smeisler** [comment]
I made some reorganizations (primarily made selecting n number of compcor components a separate function)

**Comment by pbellec** [comment]
There are a number of issues with that PR, but I am going to merge into master so I can fix them and add a test.

**Comment by pbellec** [comment]
I've created a new branch on the main repo to continue the work. I'll create and fix tests there. https://github.com/SIMEXP/load_confounds/pull/115
